src/esp/scene/SceneNode.cpp

```diff
@@ -12,26 +14,173 @@ namespace scene {
+  subtreeSensorSuite_ = new esp::sensor::SensorSuite(*this);
 }
 
-SceneNode& SceneNode::createChild() {
+SceneNode::~SceneNode() {
+  // If the entire scene graph is being deleted no need to update anything
+  if (SceneGraph::isRootNode(*this)) {
+    return;
+  }
+  // If a subtree is being deleted
+  // If parent node is nullptr already, there is no need to do anything.
+  // This is because magnum destroys the nodes in a top-down manner
+  // recursively, root node first, and then its child nodes
+  if (!this->parent()) {
+    return;
+  }
+
+  // If parent node is not nullptr, update the sensorSuites stored in each
+  // ancestor node in a bottom-up manner.
+
+  // First remove Sensor from parent's nodeSensorSuite if node is leaf node
+  removeSensorFromParentNodeSensorSuite();
+
+  // go bottom to the top, and erase sensors and all subtreeSensors from its
+  // ancestors' subtreeSensorSuites
+  removeSubtreeSensorsFromAncestors();
+}
+
+std::map<std::string, std::reference_wrapper<esp::sensor::Sensor>>&
+SceneNode::getNodeSensors() {
+  return nodeSensorSuite_->getSensors();
+}
+
+std::map<std::string, std::reference_wrapper<esp::sensor::Sensor>>&
+SceneNode::getSubtreeSensors() {
+  return subtreeSensorSuite_->getSensors();
+}
+
+SceneNode& SceneNode::createChild(SceneNodeTags childNodeTags) {
+  CORRADE_ASSERT(
+      !(sceneNodeTags_ & SceneNodeTag::Leaf),
+      "SceneNode::createChild(): Can not create child from leaf node", *this);
   // will set the parent to *this
   SceneNode* node = new SceneNode(*this);
   node->setId(this->getId());
+  node->setSceneNodeTags(childNodeTags);
   return *node;
 }
 
+SceneNode& SceneNode::setParent(SceneNode* newParent) {
+  // Perform same internal checks as magnum to ensure newParent is a valid new
+  // parent before updating any SensorSuites
+  // Parent can not be leaf node
+  CORRADE_ASSERT(!(newParent->getSceneNodeTags() & SceneNodeTag::Leaf),
+                 "SceneNode::setParent(): New parent node can not be leaf node",
+                 *this);
+
+  // Skip if this is scene or RootNode (which cannot have parent), or if
+  // newParent is already parent
+  if (isScene() || SceneGraph::isRootNode(*this) ||
+      this->parent() == newParent) {
+    return *this;
+  }
+
+  // Skip if this is an ancestor of newParent
+  auto p = newParent->parent();
+  while (p) {
+    if (p == this)
+      return *this;
+    p = p->parent();
+  }
+
+  // Remove sensors from old parent node's nodeSensorSuite
+  removeSensorFromParentNodeSensorSuite();
+
+  // Update old ancestors' SubtreeSensorSuites
+  removeSubtreeSensorsFromAncestors();
+
+  MagnumObject::setParent(newParent);
+
+  // Update new ancestors'SubtreeSensorSuites
+  addSubtreeSensorsToAncestors();
+  // Add sensors to newParent's nodeSensorSuite
+  addSensorToParentNodeSensorSuite();
+  return *this;
+}
+
+void SceneNode::addSensorToParentNodeSensorSuite() {
+  // Only add a sensor if this is a leaf node and sensor exists
```

Comment:
if this is called from the root node, and somehow the root node is marked as a leaf node, this function should just return.